components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1210,7 +1214,21 @@ where
 
         if !data.is_empty() {
             if !self.peer.is_witness || !can_witness_skip(entry) {
-                let cmd = util::parse_data_at(data, index, &self.tag);
+                let cmd = if apply_ctx.enable_v2_compatible_learner {
+                    let logger = slog_global::get_global().new(slog::o!());
+                    match SimpleWriteReqDecoder::new(
+                        |data, index, _| crate::store::util::parse_data_at(data, index, &self.tag),
+                        &logger,
+                        data,
+                        index,
+                        term,
+                    ) {
+                        Ok(simple_write_decoder) => simple_write_decoder.to_raft_cmd_request(),
+                        Err(cmd) => cmd,
+                    }
+                } else {
+                    util::parse_data_at(data, index, &self.tag)
```

Comment:
Could you make `util::parse_data_at()` returns an enum?

```rust
enum RaftCmd<'a> {
  V1(RaftCmdRequest),
  V2(SimpleWrite<'a>)
}

fn parse_data_at<'a>(data: &'a [u8], index: u64, tag: &str) -> RaftCmd<'a> { }
```

Reply:
Fixed.